src/main/java/frc/robot/subsystems/gripper/Gripper.java

```diff
@@ -0,0 +1,90 @@
+package frc.robot.subsystems.gripper;
+
+import edu.wpi.first.math.geometry.Pose3d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Rotation3d;
+import edu.wpi.first.units.Angle;
+import edu.wpi.first.units.MutableMeasure;
+import edu.wpi.first.units.Units;
+import edu.wpi.first.units.Voltage;
+import edu.wpi.first.wpilibj.smartdashboard.Mechanism2d;
+import edu.wpi.first.wpilibj.smartdashboard.MechanismLigament2d;
+import edu.wpi.first.wpilibj.smartdashboard.MechanismRoot2d;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import org.littletonrobotics.junction.AutoLogOutput;
+import org.littletonrobotics.junction.Logger;
+
+public class Gripper extends SubsystemBase {
+    private static Gripper INSTANCE;
+    private final GripperIO io;
+    private final GripperInputsAutoLogged inputs = GripperIO.inputs;
+
+    @AutoLogOutput private final Mechanism2d mechanism2d = new Mechanism2d(0, 0);
+    @AutoLogOutput private Pose3d pose3d = new Pose3d(0, 0, 0, new Rotation3d());
+
+    private final MechanismRoot2d root = mechanism2d.getRoot("Gripper", 0, 0);
+    private final MechanismLigament2d gripperLigament =
+            root.append(new MechanismLigament2d("Gripper", 0, 0));
+
+    private Gripper(GripperIO io) {
+        this.io = io;
+    }
+
+    public static Gripper getInstance() {
+        return INSTANCE;
+    }
+
+    public static void initialize(GripperIO io) {
+        INSTANCE = new Gripper(io);
+    }
+
+    public boolean hasNote() {
+        return inputs.hasNote;
+    }
+
+    public boolean atSetpoint() {
+        return inputs.atSetpoint;
+    }
+
+    public MutableMeasure<Voltage> getRollerMotorVoltage() {
+        return inputs.rollerMotorVoltage;
+    }
+
+    public MutableMeasure<Voltage> getAngleMotorVoltage() {
+        return inputs.angleMotorVoltage;
+    }
+
+    public Command setRollerPower(double power) {
+        return runOnce(() -> io.setRollerMotorPower(power)).withName("set roller power");
+    }
+
+    public Command intake() {
+        return setWristPosition(GripperConstants.INTAKE_ANGLE)
+                .alongWith(setRollerPower(GripperConstants.INTAKE_POWER))
+                .withName("intake");
+    }
+
+    public Command outtake() {
+        return setWristPosition(GripperConstants.OUTTAKE_ANGLE)
+                .alongWith(setRollerPower(GripperConstants.OUTTAKE_POWER))
+                .withName("outtake");
+    }
+
+    public Command setWristPosition(MutableMeasure<Angle> angle) {
+        return runOnce(() -> io.setAngle(angle)).withName("set wrist position");
+    }
+
+    @Override
+    public void periodic() {
+        pose3d =
+                new Pose3d(
+                        GripperConstants.GRIPPER_POSITION,
```

Comment:
Remember that the pose of the gripper depends on the height of the first stage of the elevator. So we need a way to pass this to the gripper.

Reply:
Like using elevator inputs in the gripper class?

Reply:
88691b89f6cb783836f0307304af20f246685063

Reply:
`elevatorInputs.currentHeight.mut_plus(...)` probably changes the elevator's inputs in the gripper's code which we don't want to happen. Instead of passing the whole inputs class, it's better to pass only a supplier for the elevator's height – classes should be [encapsulated](https://en.m.wikipedia.org/wiki/Encapsulation_(computer_programming)) and not export details which aren't needed. I would even make the inputs classes package-private to make sure that doesn't happen.